File: backend/services/abuseipdb_service.py
import requests
import os
from datetime import datetime
from dotenv import load_dotenv
from utils.geo_utils import get_lat_lon

load_dotenv()

class AbuseIPDBService:
    BASE_URL = "https://api.abuseipdb.com/api/v2"
    
    def __init__(self):
        self.api_key = os.getenv("ABUSEIPDB_API_KEY")
        self.headers = {
            "Key": self.api_key,
            "Accept": "application/json"
        }

    def get_blacklist(self, limit=50):
        """Fetch high-confidence blacklist IPs from AbuseIPDB."""
        if not self.api_key:
            print("ABUSEIPDB_API_KEY not set.")
            return None
            
        try:
            url = f"{self.BASE_URL}/blacklist"
            params = {"limit": limit}
            response = requests.get(url, headers=self.headers, params=params)
            response.raise_for_status()
            return response.json()
        except Exception as e:
            print(f"Error fetching blacklist from AbuseIPDB: {e}")
            return None

    def check_ip(self, ip_address):
        """Check details for a specific IP (optional, mostly for enrichment)."""
        if not self.api_key:
            return None

        try:
            url = f"{self.BASE_URL}/check"
            params = {
                "ipAddress": ip_address,
                "maxAgeInDays": 90
            }
            response = requests.get(url, headers=self.headers, params=params)
            response.raise_for_status()
            return response.json()
        except Exception as e:
            print(f"Error checking IP {ip_address}: {e}")
            return None

File: backend/services/abuseIpdbService.js
const axios = require('axios');
const path = require('path');
require('dotenv').config({ path: path.resolve(__dirname, '../.env') });

const API_KEY = process.env.ABUSEIPDB_API_KEY;
const BASE_URL = 'https://api.abuseipdb.com/api/v2';

const checkIp = async (ip) => {
    try {
        const response = await axios.get(`${BASE_URL}/check`, {
            params: {
                ipAddress: ip,
                maxAgeInDays: 90,
                verbose: true
            },
            headers: {
                'Key': API_KEY,
                'Accept': 'application/json'
            }
        });
        return response.data.data;
    } catch (error) {
        console.error('Error checking IP with AbuseIPDB:', error.message);
        return null;
    }
};

const getBlacklist = async (limit = 100) => {
    try {
        const response = await axios.get(`${BASE_URL}/blacklist`, {
            params: {
                limit: limit
            },
            headers: {
                'Key': API_KEY,
                'Accept': 'application/json'
            }
        });
        return response.data;
    } catch (error) {
        console.error('Error fetching blacklist from AbuseIPDB:', error.message);
        return [];
    }
};

module.exports = { checkIp, getBlacklist };
